File: src/adapters/llm/LLMAdapter.ts
import type { AppConfig } from '../../config/config';
import { Logger } from '../../utils/logging';

/**
 * Split markers used to identify parts of the prompt that should be preserved
 * when splitting into multiple requests
 */
export interface SplitMarkers {
  intro: string;
  outro: string;
  continuation: string;
}

/**
 * Base abstract class for LLM adapters
 * Defines the interface that all LLM adapters must implement
 */
export abstract class LLMAdapter {
  protected config: AppConfig;
  protected client?: unknown;
  protected logger: Logger;
  protected splitMarkers: SplitMarkers = {
    intro: '<!-- INTRO -->',
    outro: '<!-- OUTRO -->',
    continuation: '<!-- CONTINUATION -->',
  };

  /**
   * Create a new LLM adapter
   * @param config - The configuration for the LLM
   */
  constructor(config: AppConfig) {
    if (this.constructor === LLMAdapter) {
      throw new Error('LLMAdapter is an abstract class and cannot be instantiated directly');
    }

    this.config = config;
    this.logger = new Logger(config.verbose || false);
  }

  /**
   * Initialize the LLM client
   * @returns The initialized client
   */
  abstract initClient(): unknown;

  /**
   * Get the maximum token limit for the current provider
   * @returns The maximum number of tokens allowed in a prompt
   */
  getMaxTokenLimit(): number {
    return this.getProviderTokenLimit(this.config.llmProvider);
  }

  /**
   * Get token limit for specific provider
   * @param provider - The LLM provider name
   * @returns The token limit for the specified provider
   */
  private getProviderTokenLimit(provider: string): number {
    const providerLimits: Record<string, number> = {
      anthropic: this.config.anthropicMaxTokens,
      openai: this.config.openaiMaxTokens,
      deepseek: this.config.deepseekMaxTokens,
      copilot: this.config.copilotMaxTokens,
    };

    return providerLimits[provider] || this.config.maxPromptTokens;
  }

  /**
   * Trims whitespace from text while preserving meaningful content
   * @param text - Text to trim
   * @returns Trimmed text
   */
  trimText(text: string | undefined): string {
    if (!text) {
      return '';
    }

    // Trim leading/trailing whitespace
    let trimmed = text.trim();

    // Replace multiple consecutive whitespace with a single space
    trimmed = trimmed.replace(/\s+/g, ' ');

    // Replace multiple consecutive newlines with a single newline
    trimmed = trimmed.replace(/\n\s*\n\s*\n+/g, '\n\n');

    return trimmed;
  }

  /**
   * Split a prompt into multiple parts if it exceeds the token limit
   * @param prompt - The prompt to split
   * @returns An array of prompt parts
   */
  splitPrompt(prompt: string): string[] {
    // Check if prompt has markers and needs splitting
    if (!this.hasValidMarkers(prompt) || this.isWithinTokenLimit(prompt)) {
      return [prompt];
    }

    // Extract sections based on markers
    const { intro, outro } = this.extractMarkedSections(prompt);
    const parts = this.splitContentAtContinuationMarkers(prompt);

    // Combine parts with intro and outro
    return this.combineSplitParts(parts, intro, outro);
  }

  /**
   * Check if prompt has valid intro and outro markers
   * @param prompt - The prompt to check
   * @returns Whether prompt has valid markers
   */
  private hasValidMarkers(prompt: string): boolean {
    return prompt.includes(this.splitMarkers.intro) && prompt.includes(this.splitMarkers.outro);
  }

  /**
   * Check if prompt is within token limit
   * @param prompt - The prompt to check
   * @returns Whether prompt is within limit
   */
  private isWithinTokenLimit(prompt: string): boolean {
    return this.estimateTokenCount(prompt) <= this.getMaxTokenLimit();
  }

  /**
   * Extract intro and outro sections from prompt
   * @param prompt - The prompt with markers
   * @returns Object containing intro and outro text
   */
  private extractMarkedSections(prompt: string): { intro: string; outro: string } {
    const introMatch = prompt.match(
      new RegExp(
        `${this.splitMarkers.intro}([\\s\\S]*?)(?=${this.splitMarkers.continuation}|${this.splitMarkers.outro})`
      )
    );
    const outroMatch = prompt.match(new RegExp(`${this.splitMarkers.outro}([\\s\\S]*)$`));

    return {
      intro: introMatch ? this.trimText(introMatch[1]) : '',
      outro: outroMatch ? this.trimText(outroMatch[1]) : '',
    };
  }

  /**
   * Split the main content at continuation markers
   * @param prompt - The full prompt
   * @returns Array of content parts
   */
  private splitContentAtContinuationMarkers(prompt: string): string[] {
    // Get the main content (everything between intro and outro)
    const mainContent = prompt
      .replace(
        new RegExp(`${this.splitMarkers.intro}[\\s\\S]*?(?=${this.splitMarkers.continuation})`),
        ''
      )
      .replace(new RegExp(`${this.splitMarkers.outro}[\\s\\S]*$`), '');

    // Split main content at continuation markers
    return mainContent.split(this.splitMarkers.continuation);
  }

  /**
   * Combine split parts with intro and outro
   * @param parts - Array of content parts
   * @param intro - Intro section text
   * @param outro - Outro section text
   * @returns Array of complete prompts
   */
  private combineSplitParts(parts: string[], intro: string, outro: string): string[] {
    return parts.map((part, index) => {
      const trimmedPart = this.trimText(part);

      if (index === 0) {
        // First part gets intro + part
        return `${this.splitMarkers.intro}${intro}${trimmedPart}`;
      }
      if (index === parts.length - 1) {
        // Last part gets part + outro
        return `${trimmedPart}${this.splitMarkers.outro}${outro}`;
      }
      // Middle parts just get the part itself
      return trimmedPart;
    });
  }

  /**
   * Merge multiple responses into a single response
   * @param responses - Array of responses from LLM
   * @returns Merged response
   */
  mergeResponses(responses: string[]): string {
    // For now, just concatenate the responses
    return responses.join('\n');
  }

  /**
   * Rough estimation of token count (not perfect but good enough for safety checks)
   * @param text - The text to estimate token count for
   * @returns Estimated number of tokens
   */
  estimateTokenCount(text: string): number {
    // A simple approximation: 4 characters per token
    return Math.ceil(text.length / 4);
  }

  /**
   * Analyze code with the LLM, handling large prompts by splitting if necessary
   * @param prompt - The prompt to send to the LLM
   * @returns The response from the LLM
   */
  async analyzeCode(prompt: string): Promise<string> {
    try {
      // Apply basic trimming to the full prompt to reduce size before splitting
      const trimmedPrompt = this.applyTrimming(prompt);
      const promptParts = this.splitPrompt(trimmedPrompt);

      if (promptParts.length > 1) {
        this.logger.debug(`Prompt split into ${promptParts.length} parts due to token limit`);
        return await this.handleMultiPartRequest(promptParts);
      }

      // Standard single request
      return await this.sendRequest(promptParts[0] || '');
    } catch (error) {
      this.logger.error(`Error analyzing code: ${(error as Error).message}`);
      throw error;
    }
  }

  /**
   * Handle multiple request parts sequentially
   * @param promptParts - Array of prompt parts
   * @returns Combined response
   */
  private async handleMultiPartRequest(promptParts: string[]): Promise<string> {
    const responses: string[] = [];

    for (let i = 0; i < promptParts.length; i++) {
      this.logger.debug(`Sending part ${i + 1} of ${promptParts.length}`);
      const part = promptParts[i] || '';
      const response = await this.sendRequest(part);
      responses.push(response);
    }

    return this.mergeResponses(responses);
  }

  /**
   * Apply trimming to a prompt while preserving sections marked by special markers
   * @param prompt The prompt to trim
   * @returns The trimmed prompt with markers preserved
   */
  private applyTrimming(prompt: string): string {
    // Check if prompt has markers for splitting
    if (!this.hasValidMarkers(prompt)) {
      return this.trimText(prompt);
    }

    return this.trimWithMarkersPreserved(prompt);
  }

  /**
   * Trim text while preserving marker sections
   * @param prompt - The prompt with markers
   * @returns Trimmed prompt with preserved markers
   */
  private trimWithMarkersPreserved(prompt: string): string {
    // Extract all marker positions
    const markerPositions = this.findAllMarkerPositions(prompt);

    // Process each section between markers
    const parts = [];
    let currentIdx = 0;

    for (const marker of markerPositions) {
      // Add text before marker (trimmed)
      if (currentIdx < marker.pos) {
        const beforeMarker = prompt.substring(currentIdx, marker.pos);
        parts.push(this.trimText(beforeMarker));
      }

      // Add the marker itself
      parts.push(this.getMarkerByType(marker.type));

      // Move past the marker
      currentIdx = marker.pos + this.getMarkerByType(marker.type).length;
    }

    // Add any remaining text after the last marker
    if (currentIdx < prompt.length) {
      const afterMarkers = prompt.substring(currentIdx);
      parts.push(this.trimText(afterMarkers));
    }

    return parts.join('');
  }

  /**
   * Find all marker positions in the prompt
   * @param prompt - The prompt to search in
   * @returns Array of marker positions and types
   */
  private findAllMarkerPositions(prompt: string): Array<{ pos: number; type: string }> {
    // Find all marker positions
    const introIndex = prompt.indexOf(this.splitMarkers.intro);
    const outroIndex = prompt.indexOf(this.splitMarkers.outro);
    const continuationIndices = this.findAllContinuationMarkers(prompt);

    // Sort all marker positions
    return [
      { pos: introIndex, type: 'intro' },
      { pos: outroIndex, type: 'outro' },
      ...continuationIndices.map(pos => ({ pos, type: 'continuation' })),
    ]
      .sort((a, b) => a.pos - b.pos)
      .filter(m => m.pos !== -1);
  }

  /**
   * Find all continuation marker positions
   * @param prompt - The prompt to search in
   * @returns Array of positions
   */
  private findAllContinuationMarkers(prompt: string): number[] {
    const indices = [];
    let contIndex = prompt.indexOf(this.splitMarkers.continuation);

    while (contIndex !== -1) {
      indices.push(contIndex);
      contIndex = prompt.indexOf(this.splitMarkers.continuation, contIndex + 1);
    }

    return indices;
  }

  /**
   * Get marker string by type
   * @param type - The marker type
   * @returns Marker string
   */
  private getMarkerByType(type: string): string {
    const markerMap: Record<string, string> = {
      intro: this.splitMarkers.intro,
      outro: this.splitMarkers.outro,
      continuation: this.splitMarkers.continuation,
    };

    return markerMap[type] || '';
  }

  /**
   * Send a request to the LLM
   * @param prompt - The prompt to send
   * @returns The response from the LLM
   */
  // eslint-disable-next-line no-unused-vars
  abstract sendRequest(prompt: string): Promise<string>;

  /**
   * Process the LLM response into a consistent format
   * @param response - The raw response from the LLM
   * @returns The processed response
   */
  processResponse(response: string): string {
    // Try to extract JSON from code blocks first
    const extractedJson = this.extractJsonFromCodeBlocks(response);
    if (extractedJson) {
      return extractedJson;
    }

    // Try to parse as direct JSON
    return this.parseAsDirectJson(response);
  }

  /**
   * Extract JSON from code blocks
   * @param response - LLM response text
   * @returns JSON string or null if not valid
   */
  private extractJsonFromCodeBlocks(response: string): string | null {
    const jsonMatch = response.match(/```(?:json)?\s*([\s\S]*?)```/);
    if (!jsonMatch || !jsonMatch[1]) {
      return null;
    }

    try {
      const parsedJson = JSON.parse(jsonMatch[1].trim());
      return JSON.stringify(parsedJson);
    } catch (parseError) {
      this.logger.warn(
        `Found JSON-like content but it could not be parsed: ${(parseError as Error).message}`
      );
      return null;
    }
  }

  /**
   * Parse response as direct JSON
   * @param response - LLM response text
   * @returns Parsed JSON or original text
   */
  private parseAsDirectJson(response: string): string {
    try {
      const parsed = JSON.parse(response);
      return JSON.stringify(parsed);
    } catch (error) {
      this.logger.warn(`Response is not valid JSON, returning as plain text ${error}`);
      return response;
    }
  }
}

File: python/tiw/adapters/llm/llm_adapter.py
"""Base LLM adapter class."""

import abc
import re
from typing import Any, Dict, List, Optional, Protocol

from ...config.config import AppConfig
from ...utils.logging import Logger


class SplitMarkers:
    """Split markers used to identify parts of the prompt that should be preserved when splitting."""
    
    def __init__(self, intro: str = "<!-- INTRO -->", 
                 outro: str = "<!-- OUTRO -->", 
                 continuation: str = "<!-- CONTINUATION -->"):
        self.intro = intro
        self.outro = outro
        self.continuation = continuation


class LLMAdapter(abc.ABC):
    """Base abstract class for LLM adapters.
    
    Defines the interface that all LLM adapters must implement.
    """
    
    def __init__(self, config: AppConfig):
        """Create a new LLM adapter.
        
        Args:
            config: The configuration for the LLM
        """
        if type(self) is LLMAdapter:
            raise TypeError("LLMAdapter is an abstract class and cannot be instantiated directly")
        
        self.config = config
        self.logger = Logger(config.verbose or False)
        self.client: Any = None
        self.split_markers = SplitMarkers()
    
    @abc.abstractmethod
    def init_client(self) -> Any:
        """Initialize the LLM client.
        
        Returns:
            The initialized client
        """
        pass
    
    def get_max_token_limit(self) -> int:
        """Get the maximum token limit for the current provider.
        
        Returns:
            The maximum number of tokens allowed in a prompt
        """
        return self._get_provider_token_limit(self.config.llm_provider)
    
    def _get_provider_token_limit(self, provider: str) -> int:
        """Get token limit for specific provider.
        
        Args:
            provider: The LLM provider name
            
        Returns:
            The token limit for the specified provider
        """
        provider_limits = {
            "anthropic": self.config.anthropic_max_tokens,
            "openai": self.config.openai_max_tokens,
            "deepseek": self.config.deepseek_max_tokens,
            "copilot": self.config.copilot_max_tokens,
        }
        
        return provider_limits.get(provider, self.config.max_prompt_tokens)
    
    def trim_text(self, text: Optional[str]) -> str:
        """Trims whitespace from text while preserving meaningful content.
        
        Args:
            text: Text to trim
            
        Returns:
            Trimmed text
        """
        if not text:
            return ""
        
        # Trim leading/trailing whitespace
        trimmed = text.strip()
        
        # Replace multiple consecutive whitespace with a single space
        trimmed = re.sub(r'\s+', ' ', trimmed)
        
        # Replace multiple consecutive newlines with a single newline
        trimmed = re.sub(r'\n\s*\n\s*\n+', '\n\n', trimmed)
        
        return trimmed
    
    def split_prompt(self, prompt: str) -> List[str]:
        """Split a prompt into multiple parts if it exceeds the token limit.
        
        Args:
            prompt: The prompt to split
            
        Returns:
            An array of prompt parts
        """
        # Check if prompt has markers and needs splitting
        if not self._has_valid_markers(prompt) or self._is_within_token_limit(prompt):
            return [prompt]
        
        # Extract sections based on markers
        intro, outro = self._extract_marked_sections(prompt)
        parts = self._split_content_at_continuation_markers(prompt)
        
        # Combine parts with intro and outro
        return self._combine_split_parts(parts, intro, outro)
    
    def _has_valid_markers(self, prompt: str) -> bool:
        """Check if prompt has valid intro and outro markers.
        
        Args:
            prompt: The prompt to check
            
        Returns:
            Whether prompt has valid markers
        """
        return self.split_markers.intro in prompt and self.split_markers.outro in prompt
    
    def _is_within_token_limit(self, prompt: str) -> bool:
        """Check if prompt is within token limit.
        
        Args:
            prompt: The prompt to check
            
        Returns:
            Whether prompt is within limit
        """
        return self.estimate_token_count(prompt) <= self.get_max_token_limit()
    
    def _extract_marked_sections(self, prompt: str) -> tuple[str, str]:
        """Extract intro and outro sections from prompt.
        
        Args:
            prompt: The prompt with markers
            
        Returns:
            Object containing intro and outro text
        """
        intro_pattern = f"{re.escape(self.split_markers.intro)}([\\s\\S]*?)(?={re.escape(self.split_markers.continuation)}|{re.escape(self.split_markers.outro)})"
        outro_pattern = f"{re.escape(self.split_markers.outro)}([\\s\\S]*)$"
        
        intro_match = re.search(intro_pattern, prompt)
        outro_match = re.search(outro_pattern, prompt)
        
        intro = self.trim_text(intro_match.group(1)) if intro_match else ""
        outro = self.trim_text(outro_match.group(1)) if outro_match else ""
        
        return intro, outro
    
    def _split_content_at_continuation_markers(self, prompt: str) -> List[str]:
        """Split the main content at continuation markers.
        
        Args:
            prompt: The full prompt
            
        Returns:
            Array of content parts
        """
        # Get the main content (everything between intro and outro)
        intro_pattern = f"{re.escape(self.split_markers.intro)}[\\s\\S]*?(?={re.escape(self.split_markers.continuation)})"
        outro_pattern = f"{re.escape(self.split_markers.outro)}[\\s\\S]*$"
        
        main_content = re.sub(intro_pattern, "", prompt)
        main_content = re.sub(outro_pattern, "", main_content)
        
        # Split main content at continuation markers
        return main_content.split(self.split_markers.continuation)
    
    def _combine_split_parts(self, parts: List[str], intro: str, outro: str) -> List[str]:
        """Combine split parts with intro and outro.
        
        Args:
            parts: Array of content parts
            intro: Intro section text
            outro: Outro section text
            
        Returns:
            Array of complete prompts
        """
        combined_parts = []
        
        for i, part in enumerate(parts):
            trimmed_part = self.trim_text(part)
            
            if i == 0:
                # First part gets intro + part
                combined_parts.append(f"{self.split_markers.intro}{intro}{trimmed_part}")
            elif i == len(parts) - 1:
                # Last part gets part + outro
                combined_parts.append(f"{trimmed_part}{self.split_markers.outro}{outro}")
            else:
                # Middle parts just get the part itself
                combined_parts.append(trimmed_part)
        
        return combined_parts
    
    def merge_responses(self, responses: List[str]) -> str:
        """Merge multiple responses into a single response.
        
        Args:
            responses: Array of responses from LLM
            
        Returns:
            Merged response
        """
        # For now, just concatenate the responses
        return "\n".join(responses)
    
    def estimate_token_count(self, text: str) -> int:
        """Rough estimation of token count (not perfect but good enough for safety checks).
        
        Args:
            text: The text to estimate token count for
            
        Returns:
            Estimated number of tokens
        """
        # A simple approximation: 4 characters per token
        import math
        return math.ceil(len(text) / 4)
    
    async def analyze_code(self, prompt: str) -> str:
        """Analyze code with the LLM, handling large prompts by splitting if necessary.
        
        Args:
            prompt: The prompt to send to the LLM
            
        Returns:
            The response from the LLM
        """
        try:
            # Apply basic trimming to the full prompt to reduce size before splitting
            trimmed_prompt = self._apply_trimming(prompt)
            prompt_parts = self.split_prompt(trimmed_prompt)
            
            if len(prompt_parts) > 1:
                self.logger.debug(f"Prompt split into {len(prompt_parts)} parts due to token limit")
                return await self._handle_multi_part_request(prompt_parts)
            
            # Standard single request
            return await self.send_request(prompt_parts[0] if prompt_parts else "")
        except Exception as error:
            self.logger.error(f"Error analyzing code: {str(error)}")
            raise
    
    async def _handle_multi_part_request(self, prompt_parts: List[str]) -> str:
        """Handle multiple request parts sequentially.
        
        Args:
            prompt_parts: Array of prompt parts
            
        Returns:
            Combined response
        """
        responses = []
        
        for i, part in enumerate(prompt_parts):
            self.logger.debug(f"Sending part {i + 1} of {len(prompt_parts)}")
            response = await self.send_request(part or "")
            responses.append(response)
        
        return self.merge_responses(responses)
    
    def _apply_trimming(self, prompt: str) -> str:
        """Apply trimming to a prompt while preserving sections marked by special markers.
        
        Args:
            prompt: The prompt to trim
            
        Returns:
            The trimmed prompt with markers preserved
        """
        # Check if prompt has markers for splitting
        if not self._has_valid_markers(prompt):
            return self.trim_text(prompt)
        
        return self._trim_with_markers_preserved(prompt)
    
    def _trim_with_markers_preserved(self, prompt: str) -> str:
        """Trim text while preserving marker sections.
        
        Args:
            prompt: The prompt with markers
            
        Returns:
            Trimmed prompt with preserved markers
        """
        # Extract all marker positions
        marker_positions = self._find_all_marker_positions(prompt)
        
        # Process each section between markers
        parts = []
        current_idx = 0
        
        for marker in marker_positions:
            # Add text before marker (trimmed)
            if current_idx < marker["pos"]:
                before_marker = prompt[current_idx:marker["pos"]]
                parts.append(self.trim_text(before_marker))
            
            # Add the marker itself
            parts.append(self._get_marker_by_type(marker["type"]))
            
            # Move past the marker
            current_idx = marker["pos"] + len(self._get_marker_by_type(marker["type"]))
        
        # Add any remaining text after the last marker
        if current_idx < len(prompt):
            after_markers = prompt[current_idx:]
            parts.append(self.trim_text(after_markers))
        
        return "".join(parts)
    
    def _find_all_marker_positions(self, prompt: str) -> List[Dict[str, Any]]:
        """Find all marker positions in the prompt.
        
        Args:
            prompt: The prompt to search in
            
        Returns:
            Array of marker positions and types
        """
        # Find all marker positions
        intro_index = prompt.find(self.split_markers.intro)
        outro_index = prompt.find(self.split_markers.outro)
        continuation_indices = self._find_all_continuation_markers(prompt)
        
        # Sort all marker positions
        all_markers = [
            {"pos": intro_index, "type": "intro"},
            {"pos": outro_index, "type": "outro"},
        ]
        
        all_markers.extend([{"pos": pos, "type": "continuation"} for pos in continuation_indices])
        
        # Sort by position and filter out non-existent markers
        return sorted(
            [m for m in all_markers if m["pos"] != -1],
            key=lambda x: x["pos"]
        )
    
    def _find_all_continuation_markers(self, prompt: str) -> List[int]:
        """Find all continuation marker positions.
        
        Args:
            prompt: The prompt to search in
            
        Returns:
            Array of positions
        """
        indices = []
        idx = prompt.find(self.split_markers.continuation)
        
        while idx != -1:
            indices.append(idx)
            idx = prompt.find(self.split_markers.continuation, idx + 1)
        
        return indices
    
    def _get_marker_by_type(self, marker_type: str) -> str:
        """Get marker string by type.
        
        Args:
            type: The marker type
            
        Returns:
            Marker string
        """
        marker_map = {
            "intro": self.split_markers.intro,
            "outro": self.split_markers.outro,
            "continuation": self.split_markers.continuation,
        }
        
        return marker_map.get(marker_type, "")
    
    @abc.abstractmethod
    async def send_request(self, prompt: str) -> str:
        """Send a request to the LLM.
        
        Args:
            prompt: The prompt to send
            
        Returns:
            The response from the LLM
        """
        pass
    
    def process_response(self, response: str) -> str:
        """Process the LLM response into a consistent format.
        
        Args:
            response: The raw response from the LLM
            
        Returns:
            The processed response
        """
        # Try to extract JSON from code blocks first
        extracted_json = self._extract_json_from_code_blocks(response)
        if extracted_json:
            return extracted_json
        
        # Try to parse as direct JSON
        return self._parse_as_direct_json(response)
    
    def _extract_json_from_code_blocks(self, response: str) -> Optional[str]:
        """Extract JSON from code blocks.
        
        Args:
            response: LLM response text
            
        Returns:
            JSON string or None if not valid
        """
        json_match = re.search(r'```(?:json)?\s*([\s\S]*?)```', response)
        if not json_match or not json_match.group(1):
            return None
        
        try:
            import json
            parsed_json = json.loads(json_match.group(1).strip())
            return json.dumps(parsed_json)
        except Exception as parse_error:
            self.logger.warn(f"Found JSON-like content but it could not be parsed: {str(parse_error)}")
            return None
    
    def _parse_as_direct_json(self, response: str) -> str:
        """Parse response as direct JSON.
        
        Args:
            response: LLM response text
            
        Returns:
            Parsed JSON or original text
        """
        try:
            import json
            parsed = json.loads(response)
            return json.dumps(parsed)
        except Exception as error:
            self.logger.warn(f"Response is not valid JSON, returning as plain text {error}")
            return response